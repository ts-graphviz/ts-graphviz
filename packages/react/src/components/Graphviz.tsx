import React, { FunctionComponent } from 'react';
import { GraphvizContext } from './GraphvizContext';

export const Graphviz: FunctionComponent = ({ children }) => {
  return <GraphvizContext.Provider value={{}}>{children}</GraphvizContext.Provider>;
};
