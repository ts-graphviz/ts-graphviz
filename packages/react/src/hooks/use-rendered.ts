import { useState, useEffect, useMemo } from 'react';

export type Format = 'svg' | 'dot' | 'json' | 'dot_json' | 'xdot_json';

export type Engine = 'circo' | 'dot' | 'fdp' | 'neato' | 'osage' | 'patchwork' | 'twopi';
export type Image = {
  path: string;
  width: string;
  height: string;
};
export type File = {
  path: string;
  data: string;
};

export const useRendered = (
  dot: string,
  engine?: Engine,
  format?: Format,
  ext?: {
    images?: Image[];
    files?: File[];
  },
): string | undefined => {
  const [rendered, setRendered] = useState<string>();
  const graphviz = useMemo(() => import('@hpcc-js/wasm').then(wasm => wasm.graphviz), []);
  useEffect(() => {
    graphviz.then(gv => gv.layout(dot, format, engine, ext)).then(setRendered);
  }, [dot, engine, format, ext, setRendered, graphviz]);
  return rendered;
};
