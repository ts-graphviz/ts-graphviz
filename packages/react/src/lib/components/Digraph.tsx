import { type FC, useEffect } from 'react';
import { CurrentGraph } from '../contexts/CurrentGraph.js';
import { GraphContainer } from '../contexts/GraphContainer.js';
import { useDigraph } from '../hooks/useDigraph.js';
import { useGraphContainer } from '../hooks/useGraphContainer.js';
import { useGraphMap } from '../hooks/useGraphMap.js';
import { useRenderedID } from '../hooks/useRenderedID.js';
import type { RootGraphProps } from '../types.js';

/**
 * `Digraph` component.
 */
export const Digraph: FC<RootGraphProps> = ({
  children,
  label,
  ...options
}) => {
  const container = useGraphContainer();
  if (container !== null) {
    throw Error(
      'RootCluster is duplicated.\nUse only one of Digraph and Graph.',
    );
  }
  const renderedLabel = useRenderedID(label);
  if (renderedLabel !== undefined)
    Object.assign(options, { label: renderedLabel });
  const digraph = useDigraph(options);
  const clusters = useGraphMap();
  useEffect(() => {
    if (digraph.id !== undefined) {
      clusters.set(digraph.id, digraph);
    }
  }, [clusters, digraph]);
  return (
    <GraphContainer.Provider value={digraph}>
      <CurrentGraph.Provider value={digraph}>{children}</CurrentGraph.Provider>
    </GraphContainer.Provider>
  );
};

Digraph.displayName = 'Digraph';
