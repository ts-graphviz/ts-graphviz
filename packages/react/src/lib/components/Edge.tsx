import type { FC } from 'react';
import { useEdge } from '../hooks/useEdge.js';
import { useRenderedID } from '../hooks/useRenderedID.js';
import type { EdgeProps } from '../types.js';

/**
 * `Edge` component.
 */
export const Edge: FC<EdgeProps> = ({ targets, label, ...options }) => {
  const renderedLabel = useRenderedID(label);
  if (renderedLabel !== undefined)
    Object.assign(options, { label: renderedLabel });
  useEdge(targets, options);
  return null;
};

Edge.displayName = 'Edge';
