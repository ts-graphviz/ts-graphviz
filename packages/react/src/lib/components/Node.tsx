import type { FC } from 'react';
import { useNode } from '../hooks/useNode.js';
import { useRenderedID } from '../hooks/useRenderedID.js';
import type { NodeProps } from '../types.js';

/**
 * `Node` component.
 */
export const Node: FC<NodeProps> = ({ id, label, xlabel, ...options }) => {
  const renderedLabel = useRenderedID(label);
  const renderedXlabel = useRenderedID(xlabel);

  if (renderedLabel !== undefined)
    Object.assign(options, { label: renderedLabel });
  if (renderedXlabel !== undefined)
    Object.assign(options, { xlabel: renderedXlabel });
  useNode(id, options);
  return null;
};

Node.displayName = 'Node';
