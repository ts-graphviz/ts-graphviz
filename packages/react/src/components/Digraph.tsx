import React, { FC, useMemo } from 'react';
import PropTypes from 'prop-types';
import * as gv from 'ts-graphviz';
import { RootClusterContext } from '../contexts/RootClusterContext';
import { ClusterContext } from '../contexts/ClusterContext';
import { useGraphvizContext } from '../hooks/useContext';

type Props = {
  id?: string;
};

export const Digraph: FC<Props> = ({ children, id }) => {
  const context = useGraphvizContext();
  const g = useMemo(() => new gv.Digraph(context, id), [context, id]);
  return (
    <RootClusterContext.Provider value={g}>
      <ClusterContext.Provider value={g}>{children}</ClusterContext.Provider>
    </RootClusterContext.Provider>
  );
};

Digraph.displayName = 'Digraph';

Digraph.defaultProps = {
  id: undefined,
};

Digraph.propTypes = {
  id: PropTypes.string,
};
