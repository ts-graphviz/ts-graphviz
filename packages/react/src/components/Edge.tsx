import React, { FC, useEffect, useMemo, ReactElement, isValidElement } from 'react';
import PropTypes from 'prop-types';
import { EdgeAttributes, EdgeTargetLike } from 'ts-graphviz';
import { renderToStaticMarkup } from 'react-dom/server';
import { useCluster } from '../hooks/useCluster';

type ReactEdgeAttributes = Omit<EdgeAttributes, 'label'> & {
  label?: ReactElement | string;
};

function renderId(id?: ReactElement | string): string | undefined {
  if (typeof id === 'string') {
    return id;
  }
  if (isValidElement(id)) {
    return renderToStaticMarkup(id);
  }
  return undefined;
}

type Props = {
  targets: EdgeTargetLike[];
  comment?: string;
} & ReactEdgeAttributes;

export const Edge: FC<Props> = ({ children, targets, comment, ...attributes }) => {
  const cluster = useCluster();
  const edge = useMemo(() => cluster.createEdge(...targets), [cluster, targets]);
  useEffect(() => {
    const { label, ...rest } = attributes;
    edge.attributes.clear();
    edge.attributes.apply({ label: renderId(label), ...rest });
  }, [edge, attributes]);

  useEffect(() => {
    edge.comment = comment;
  }, [edge, comment]);

  useEffect(() => {
    return (): void => {
      cluster.removeEdge(edge);
    };
  }, [cluster, edge]);
  return <>{children}</>;
};

Edge.displayName = 'Edge';

Edge.propTypes = {
  targets: PropTypes.array.isRequired,
  comment: PropTypes.string,
  label: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
};

Edge.defaultProps = {
  comment: undefined,
  label: undefined,
};
