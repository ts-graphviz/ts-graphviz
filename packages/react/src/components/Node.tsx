import React, { FC, ReactElement } from 'react';
import PropTypes from 'prop-types';
import { useNode, NodeProps } from '../hooks/use-node';
import { useRenderedID } from '../hooks/use-rendered-id';

type Props = Omit<NodeProps, 'label'> & {
  label?: ReactElement | string;
};

export const Node: FC<Props> = ({ children, label, ...props }) => {
  const renderedLabel = useRenderedID(label);
  if (renderedLabel !== undefined) Object.assign(props, { label: renderedLabel });
  useNode(props);
  return <>{children}</>;
};

Node.displayName = 'Node';

Node.propTypes = {
  id: PropTypes.string.isRequired,
  comment: PropTypes.string,
  label: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
};

Node.defaultProps = {
  comment: undefined,
  label: undefined,
};
