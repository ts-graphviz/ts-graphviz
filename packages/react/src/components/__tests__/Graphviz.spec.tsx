import React from 'react';
import { render } from '@testing-library/react';
import { Graphviz } from '../Graphviz';
import { Engine, Format, Image } from '../../hooks/use-rendered';
import { Digraph } from '../Digraph';
import { Node } from '../Node';
import { Edge } from '../Edge';

jest.mock('../../hooks/use-rendered', () => {
  return {
    useRendered: (
      dot: string,
      engine?: Engine,
      format?: Format,
      ext?: {
        images?: Image[];
        files?: File[];
      },
    ): string => JSON.stringify({ dot, engine, format, ext }),
  };
});

describe('Graphviz component', () => {
  test('render dot element', () => {
    const { container } = render(
      <Graphviz>
        <Digraph>
          <Node id="n1" />
          <Node id="n2" />
          <Node id="n3" />
          <Edge targets={['n1', 'n2', 'n3']} />
          <Edge targets={['n1', 'n3']} />
        </Digraph>
      </Graphviz>,
    );
    expect(container).toMatchSnapshot();
  });
});
