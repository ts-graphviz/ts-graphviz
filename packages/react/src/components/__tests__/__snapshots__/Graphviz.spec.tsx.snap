// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graphviz component render dot element 1`] = `
<div>
  <div>
    {"dot":"digraph {\\n  \\"n1\\";\\n  \\"n2\\";\\n  \\"n3\\";\\n  \\"n1\\" -&gt; \\"n2\\" -&gt; \\"n3\\";\\n  \\"n1\\" -&gt; \\"n3\\";\\n}","format":"svg","ext":{"images":[],"files":[]}}
  </div>
</div>
`;
